state/remoteapplication.go

```diff
@@ -290,6 +290,91 @@ func (s *RemoteApplication) Endpoint(relationName string) (Endpoint, error) {
 	return Endpoint{}, fmt.Errorf("remote application %q has no %q relation", s, relationName)
 }
 
+// AddEndpoints adds the specified endpoints to the remote application.
+// If an endpoint already exists, the new endpoint needs to be identical
+// or else an error is returned.
+func (s *RemoteApplication) AddEndpoints(eps []charm.Relation) error {
+	newEps := make([]remoteEndpointDoc, len(eps))
+	for i, ep := range eps {
+		newEps[i] = remoteEndpointDoc{
+			Name:      ep.Name,
+			Role:      ep.Role,
+			Interface: ep.Interface,
+			Limit:     ep.Limit,
+			Scope:     ep.Scope,
+		}
+	}
+
+	model, err := s.st.Model()
+	if err != nil {
+		return errors.Trace(err)
+	} else if model.Life() != Alive {
+		return errors.Errorf("model is no longer alive")
+	}
+
+	checkCompatibleEndpoints := func(currentEndpoints []Endpoint) error {
+		// Ensure we don't have any existing endpoints that are
+		// not identical with the new ones.
+		currentEndpointsMap := make(map[string]Endpoint)
+		for _, ep := range currentEndpoints {
+			currentEndpointsMap[ep.Name] = ep
+		}
+		for _, r := range eps {
+			existing, ok := currentEndpointsMap[r.Name]
+			if !ok {
+				continue
+			}
+			if existing.Relation != r {
+				return errors.Errorf("conflicting endpoint %v", r.Name)
+			}
+		}
+		return nil
+	}
+
+	currentEndpoints, err := s.Endpoints()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if err := checkCompatibleEndpoints(currentEndpoints); err != nil {
+		return err
+	}
+	applicationID := s.st.docID(s.Name())
+	buildTxn := func(attempt int) ([]txn.Op, error) {
+		// If we've tried once already and failed, check that
+		// model may have been destroyed.
+		if attempt > 0 {
+			if err := checkModelActive(s.st); err != nil {
+				return nil, errors.Trace(err)
+			}
+			currentEndpoints, err = s.Endpoints()
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			if err := checkCompatibleEndpoints(currentEndpoints); err != nil {
+				return nil, err
+			}
+		}
+		ops := []txn.Op{
+			model.assertActiveOp(),
+			{
+				C:  remoteApplicationsC,
+				Id: applicationID,
+				Assert: bson.D{
+					{"endpoints", bson.D{{"$size", len(currentEndpoints)}}},
```

Comment:
probably better to check the set matches exactly? we could conceivably make changes to allow endpoints to be removed, when the remote application is not involved in any relations that use the existing endpoints

if you take my previous suggestion of always rejecting existing endpoints, then you could instead assert that the existing endpoints don't exist; we wouldn't care about addition/removal of other endpoints